tests/utils/devServer.js

```diff
@@ -0,0 +1,92 @@
+const cliPath = require('./cliPath')
+const path = require('path')
+const getPort = require('get-port')
+const seedrandom = require('seedrandom')
+const execa = require('execa')
+const pidtree = require('pidtree')
+
+// each process gets a starting port based on the pid
+const rng = seedrandom(`${process.pid}`)
+function getRandomPortStart(rng) {
+  const startPort = Math.floor(rng() * 10000) + 10000 // 10000 to avoid collisions with frameworks ports
+  return startPort
+}
+
+let currentPort = getRandomPortStart(rng)
+
+const startServer = async ({ cwd, env = {} }) => {
+  const tryPort = currentPort++
+  const port = await getPort({ port: tryPort })
+  const host = 'localhost'
+  const url = `http://${host}:${port}`
+  console.log(`Starting dev server on port: ${port} in directory ${path.basename(cwd)}`)
+  const ps = execa(cliPath, ['dev', '-p', port], {
+    cwd,
+    env: { ...process.env, ...env },
+    reject: false,
+  })
+  return new Promise((resolve, reject) => {
+    ps.stdout.on('data', data => {
+      if (data.toString().includes('Server now ready on')) {
+        resolve({
+          url,
+          host,
+          port,
+          close: async () => {
+            const pids = await pidtree(ps.pid).catch(() => [])
+            pids.forEach(pid => () => {
+              try {
+                process.kill(pid)
+              } catch (e) {
+                // no-op
+              }
+            })
+            ps.kill()
+            await Promise.race([ps, new Promise(resolve => setTimeout(resolve, 1000))])
+          },
+        })
+      }
+    })
+
+    let error = ''
+    ps.stderr.on('data', data => {
+      error = error + data.toString()
+    })
+    ps.on('close', code => {
```

Comment:
You might want to look for `ps` rejection instead (`ps` is a promise as well). A child process can fail for more reasons than exit code non-0:
  - signal termination (exit code will be `undefined`)
  - `error` event on `stdin`, `stdout` or `stderr`
  - child spawning error

Reply:
That's a good point.
We can't `await` on `ps` since the dev process never exits until we kill it (unless there was an unexpected error).
The code here is mostly intended to catch errors with the dev server not initialising properly (and prevent waiting indefinitely for the `Server now ready on` message).

Reply:
We could do something like: 

```js
ps.catch(error => {
  reject(error)
})
```

Reply:
Fixed in https://github.com/netlify/cli/pull/1040/commits/0f0d6bdb0c2a10dcde6e2b51cdfdfeadc333a9f2